Distinguish automatically disabled checkers

fmdkdd opened the pull request:
This commit does the following in order to resolve #1542:

1. Add a variable `-automatically-disabled-checkers` which distinguish
between manually disabled checkers and automatically disabled ones.
This allows us to reset the automatically disabled status without
overwriting user preference from `-disabled-checkers`.

2. Add a function `-reset-enabled-checker` to reset the enabled
or automatically disabled status of a checker.  This allows us to retest
the `:enabled` predicate for a checker.  This function is also useful
for external plugins that may have knowledge that circumstances that led
to a failed `:enabled` test have changed, and a new check must be
done (see #1542).

3. Call `-reset-enabled-checker` in `-verify-checker` so that manually
triggering a verification checks the enabled state again.

Closes #1542.

fmdkdd commented:
Sending for reviewing @cpitclaudel.  If we agree on the specifics, I will add tests before merging.

cpitclaudel commented on flycheck.el:
```diff
@@ -312,6 +312,22 @@ respectively."
+The command `flycheck-verify-setup' will clear this variable and
+reverify all checkers for the current buffer.  Checkers that fail
+the `:enabled` test will remain automatically disabled.  Checkers
+that pass will be enabled.
```
I think this is a bit dangerous: don't we fear that we'll have heisenbugs were something doesn't work, we ask for the contents of `verify`, and and the bugs go away?

fmdkdd replied:
That is risk indeed.  This convenience may in fact mystify users.  Better to have the user trigger the reverification (using buttons like in #1478).

Besides, the reverification can delay opening the verify-setup buffer, as in js-mode where the enabled test for `eslint` takes its sweet time.

I'll remove this auto-reset and we'll deal with the interaction bit in #1478.

cpitclaudel commented on flycheck.el:
```diff
@@ -312,6 +312,22 @@ respectively."
   :safe #'flycheck-symbol-list-p)
 (make-variable-buffer-local 'flycheck-disabled-checkers)
 
+(defvar-local flycheck-automatically-disabled-checkers nil
```
`flycheck--` instead of `flycheck-`, maybe?

fmdkdd replied:
I forget we are using this convention now.  :+1:

cpitclaudel commented on flycheck.el:
```diff
@@ -1556,7 +1572,16 @@ A checker is registered if it is contained in
+
+A checker has been automatically disabled if it is contained in
+`flycheck-automatically-disabled-checkers'."
+  (memq checker flycheck-automatically-disabled-checkers))
```
Do you want to use that function in flycheck-disabled-checker-p?

fmdkdd replied:
Good catch.

cpitclaudel commented on flycheck.el:
```diff
@@ -1945,9 +1970,26 @@ nil otherwise."
+always be constant (t or nil respectively).
+
+If you wish to test the `:enabled' predicate again, you must
+first reset its state using this function."
```
I think this description is a bit too technical.  Maybe just something along the lines of `clear Flycheck's cache of :enabled properties`?

EDIT: Oh, except I see that this removes the value from `flycheck-disabled-checkers`, too.  But see below

cpitclaudel commented on flycheck.el:
```diff
@@ -1945,9 +1970,26 @@ nil otherwise."
+          (remq checker flycheck-automatically-disabled-checkers)))
+  (when (memq checker flycheck-enabled-checkers)
+    (setq flycheck-enabled-checkers
+          (remq checker flycheck-enabled-checkers))))
```
I think we don't want that, since we call this function from `flycheck-verify-checker`

fmdkdd replied:
That's the point though: if a checker has previously passed the `:enabled` test, it has been added to `flycheck-enabled-checkers` and won't be tested again unless we remove it from that list.

Resetting the `:enabled` state thus means removing the checker from these two lists.

cpitclaudel replied:
Oh, you're completely right. I read `disabled` instead of `enabled` here.

cpitclaudel commented on flycheck.el:
```diff
@@ -2114,7 +2156,7 @@ Return a list of `flycheck-verification-result' objects."
         (verify (flycheck-checker-get checker 'verify)))
     (when enabled
-      (let ((result (funcall enabled)))
+      (let ((result (flycheck-may-enable-checker checker)))
```
Can you explain this part? (IIUC, this will return the same result, since we'll have reset the cache in the meantime?)

fmdkdd replied:
`-verify-checker` ended up calling the `:enabled` predicate twice: once through `verify-generic-checker`, and once through `may-use-checker`.

Because `-may-enable` tests `:enabled` *and* caches its result, using `-may-enable` here ensures we run the predicate only once.

The `(funcall enabled)` above was obviously bypassing the cached result.

(Note that `-verify-checker` is seldom used though)

cpitclaudel commented on flycheck.el:
```diff
@@ -1945,9 +1970,26 @@ nil otherwise."
                                 (funcall enabled)))))
     (if shall-enable
         (cl-pushnew checker flycheck-enabled-checkers)
```
Don't we want to rename this to `flycheck-automatically-enabled-checkers`, for symmetry? (We'll need an alias, since https://github.com/search?q=flycheck-enabled-checkers&type=Code some people seem to use that variable?!)

fmdkdd replied:
Yes, I agree that it makes sense for the symmetry.  I didn't do it precisely because I didn't want to break the API.

Maybe we add an alias and deprecate the former one?

cpitclaudel replied:
An alias sounds good.

fmdkdd commented:
So apparently `defvaralias` can't handle local variables:
```elisp
(define-obsolete-variable-alias 'flycheck-enabled-checkers
  'flycheck-automatically-enabled-checkers "32")

-> (error "Don’t know how to make a localized variable an alias")
```
But, strangely, we can define the alias before making it buffer-local, and it seems to work:
```elisp
(defvar test-local-var 1)
(define-obsolete-variable-alias 'obsolete-var 'test-local-var)
(make-variable-buffer-local 'test-local-var)
(setq obsolete-var 2)
test-local-var -> 2
In other buffer:
test-local-var -> 1
```
Shall we go ahead with the workaround?

cpitclaudel commented:
Yes, a quick look at the C sources suggests that this should be fine.

fmdkdd commented:
I've added the alias, renamed occurences of the obsolete variable, and fixed a bug with `flycheck-disable-checker` that was unable to enable checkers that were automatically disabled.

fmdkdd commented:
Oh and I've also extended the error message for excessive checkers, as suggested in #748.  And added `verify-setup` to the modemap (and menubar) menu, to help discovery.

fmdkdd commented:
Didn't see #1580, let's go with you error message instead, it's clearer.
